task_registry/data.py

```diff
@@ -1,67 +1,117 @@
 import logging
 import os
+from enum import StrEnum
 from pathlib import Path
 from typing import Any
 
 import yaml
+from pydantic import BaseModel, Field
 
 logger = logging.getLogger(__name__)
 
 
-def get_file_size(file_path: str) -> int:
-    return os.path.getsize(file_path)  # pragma: no cover
+class Tasks(StrEnum):
+    INSTRUMENTS = "instruments"
+    REQUIREMENTS = "requirements"
+    MEASURES = "measures"
 
 
-def create_urn_mappper(entries: list[dict[str, Any]]) -> dict[str, Any]:
-    urn_mapper: dict[str, Any] = {}
-    for instrument in entries:
-        path = Path(instrument["path"])
-        try:
-            with open(str(path)) as f:
-                urn_mapper[instrument["urn"]] = yaml.safe_load(f)
-        except FileNotFoundError:
-            logger.exception(f"Instrument file with path {path} not found.")  # pragma: no cover
+class Link(BaseModel):
+    self: str
 
-        except yaml.YAMLError:
-            logger.exception(f"Instrument file with path {path} could not be parsed.")  # pragma: no cover
 
-    return urn_mapper
+class FileInfo(BaseModel):
+    type: str
+    size: int
+    name: str
+    path: str
+    urn: str
+    download_url: str
+    links: Link
+
+
+class Index(BaseModel):
+    type: str = Field(examples=["dir"])
+    size: int = Field(examples=[0])
+    name: str = Field(examples=["task_collection_name"])
+    path: str = Field(examples=["task_collection_path"])
+    download_url: str = Field(examples=["https://task-registry.apps.digilab.network/task_collection"])
+    links: Link = Field(examples=[{"self": "https://task-registry.apps.digilab.network"}])
+    entries: list[FileInfo] = Field(
+        examples=[
+            {
+                "type": "file",
+                "size": 1024,
+                "name": "task_name.yaml",
+                "path": "task_collection_path/task_name.yaml",
+                "urn": "urn:nl:aivt:tr:xx:xx",
+                "download_url": "https://task-registry.apps.digilab.network/task_collection/urn/urn:nl:aivt:tr:aiia:1.0",
+                "links": {
+                    "self": "https://task-registry.apps.digilab.network/task_collection/urn/urn:nl:aivt:tr:aiia:1.0"
+                },
+            }
+        ]
+    )
+
+
+class CachedRegistry:
+    def __init__(self) -> None:
+        self.index_cache: dict[Tasks, Index] = {}
+        self.tasks_cache: dict[tuple[str, Tasks], Any] = {}
+
+    def add_tasks(self, tasks: Tasks) -> None:
+        index = generate_index(tasks)
+        self.index_cache[tasks] = index
+
+        for task in index.entries:
+            path = Path(task.path)
+            try:
+                with open(str(path)) as f:
```

Comment:
You first make a Path() from a string (?) and then use the str of the Path again in the open... do you need Path() at all then?